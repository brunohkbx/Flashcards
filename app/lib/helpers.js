import { Platform } from 'react-native';

export const waitFor = ms => {
  return new Promise(resolve => setTimeout(resolve, ms));
};

export const getOS = () => {
  return Platform.OS;
};
