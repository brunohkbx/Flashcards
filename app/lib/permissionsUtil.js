import { Permissions } from 'expo';

export const obtainNotificationPermission = async () => {
  const { status: existingStatus } = await Permissions.getAsync(
    Permissions.NOTIFICATIONS
  );
  let finalStatus = existingStatus;

  if (existingStatus !== 'granted') {
    const { status } = await Permissions.askAsync(Permissions.NOTIFICATIONS);

    finalStatus = status;
  }

  return finalStatus;
};
