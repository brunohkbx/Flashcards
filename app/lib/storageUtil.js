import { AsyncStorage } from 'react-native';
import { pickBy, assign } from 'lodash/object';

const DECK_STORAGE_KEY = 'flashcards:deck';

export const getDecks = () => {
  return AsyncStorage.getItem(DECK_STORAGE_KEY);
}

export const createDeck = deck => {
  return AsyncStorage.mergeItem(DECK_STORAGE_KEY, JSON.stringify(deck));
}

export const deleteDeck = id => {
  return getDecks()
    .then(decks => JSON.parse(decks))
    .then(decks => pickBy(decks, deck => deck.id !== id))
    .then(decks => AsyncStorage.setItem(DECK_STORAGE_KEY, JSON.stringify(decks)));
}

export const editDeck = deck => {
  return getDecks()
    .then(decks => JSON.parse(decks))
    .then(decks => assign(decks, deck))
    .then(decks => AsyncStorage.setItem(
      DECK_STORAGE_KEY, JSON.stringify(decks)
    ));
}
