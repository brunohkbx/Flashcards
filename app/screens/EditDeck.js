import React, { Component } from 'react';
import { connect } from 'react-redux';
import Container from '../components/Container';
import DeckForm from '../components/DeckForm';
import { editDeck } from '../actions';
import { Colors, Appbar } from 'react-native-paper';

export class EditDeck extends Component {
  static navigationOptions = ({ navigation }) => {
    return {
      headerRight:
        <Appbar.Action
          icon="done"
          color={Colors.white}
          onPress={() => navigation.getParam('submitForm')()}
        />
    };
  };

  componentDidMount() {
    this.props.navigation.setParams({ submitForm: this.submitForm });
  }

  handleSubmit = (values, actions) => {
    const { editDeck, navigation } = this.props;

    editDeck(values);
    actions.setSubmitting(false);
    navigation.navigate(
      'Main',
      { flashMessage: 'Deck has been successfully edited' }
    );
  }

  submitForm = () => this.form.submitForm();

  render() {
    const { deck } = this.props;

    return (
      <Container>
        <DeckForm
          initialValues={deck}
          formRef={node => this.form = node}
          handleSubmit={this.handleSubmit}
        />
      </Container>
    );
  }
}

const mapStateToProps = ({ decks }, { navigation }) => {
  const { deckId } = navigation.state.params;

  return {
    deck: decks[deckId]
  };
};

const mapDispatchToProps = dispatch => ({
  editDeck: data => dispatch(editDeck(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditDeck);
