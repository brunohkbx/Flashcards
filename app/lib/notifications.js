import { Notifications } from 'expo';
import * as storageUtil from '../lib/storageUtil';

export const clearLocalNotification = () => {
  return storageUtil.unscheduleNotification()
    .then(Notifications.cancelAllScheduledNotificationsAsync);
};

export const scheduleLocalNotification = () => {
  return storageUtil.isScheduledNotification()
    .then(JSON.parse)
    .then(data => {
      if (data === null) {
        Notifications.cancelAllScheduledNotificationsAsync();

        let tomorrow = new Date();
        tomorrow.setDate(tomorrow.getDate() + 1);
        tomorrow.setHours(18);
        tomorrow.setMinutes(30);

        Notifications.scheduleLocalNotificationAsync(
          {
            title: 'Haven\'t seen you in a while',
            body: 'ðŸ‘‹ Don\'t forget to take your quiz today!',
            ios: {
              sound: true
            },
            android: {
              sound: true,
              priority: 'high',
              sticky: false,
              vibrate: true
            }
          },
          {
            time: tomorrow,
            repeat: 'day',
          }
        );

        storageUtil.scheduleNotification();
      }
    });
};
