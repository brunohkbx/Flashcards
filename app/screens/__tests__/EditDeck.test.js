import React from 'react';
import { shallow } from 'enzyme';
import { EditDeck } from '../EditDeck';
import * as helpers from '../../lib/helpers';

describe('EditDeck', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      navigation: {
        setParams: jest.fn(),
        navigate: jest.fn()
      },
      editDeck: jest.fn(),
      deck: { id: '606a1255', questions: [], title: 'React' }
    }, propOverrides)

    const wrapper = shallow(<EditDeck {...defaultProps} />);
    const wrapperInstance = wrapper.instance();

    return {
      wrapper,
      wrapperInstance
    }
  }

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('assigns submitForm to navigations params when component did mount', () => {
    const mockNavigation = { setParams: jest.fn() }
    const { wrapperInstance } = setup({ navigation: mockNavigation });

    expect(
      mockNavigation.setParams
    ).toHaveBeenCalledWith({ submitForm: wrapperInstance.submitForm });
  });

  describe('handleSubmit', () => {
    it('displays the progressDialog', async () => {
      const mockEditDeck = jest.fn(() => Promise.resolve());
      const { wrapperInstance } = setup({ editDeck: mockEditDeck });
      jest.spyOn(wrapperInstance, 'setState');

      await wrapperInstance.handleSubmit({}, {});

      expect(
        wrapperInstance.setState
      ).toHaveBeenCalledWith({ progressDialogVisible: true });
    });

    it('waits for 1000 ms and then calls submitEditedDeck', async () => {
      const mockValues = jest.fn();
      const mockEditDeck = jest.fn(() => Promise.resolve());
      const { wrapperInstance } = setup({ editDeck: mockEditDeck });
      jest.spyOn(helpers, 'waitFor');
      jest.spyOn(wrapperInstance, 'submitEditedDeck');

      await wrapperInstance.handleSubmit(mockValues, {});

      expect(helpers.waitFor).toHaveBeenCalledWith(1000);
      expect(wrapperInstance.submitEditedDeck).toHaveBeenCalledWith(mockValues);
    });
  });

  describe('submitForm', () => {
    it('calls submitForm on formik ref', () => {
      const mockSubmitForm = jest.fn();
      const { wrapperInstance } = setup();
      wrapperInstance.form = { submitForm: mockSubmitForm };

      wrapperInstance.submitForm();

      expect(mockSubmitForm).toHaveBeenCalled();
    });
  });

  describe('submitEditedDeck', () => {
    it('calls editDeck with the new values', () => {
      const mockValues = jest.fn();
      const mockEditDeck = jest.fn(() => Promise.resolve());
      const { wrapperInstance } = setup({ editDeck: mockEditDeck });

      wrapperInstance.submitEditedDeck(mockValues);

      expect(mockEditDeck).toHaveBeenCalledWith(mockValues);
    });

    it('then hides the progressDialog', async () => {
      const mockEditDeck = jest.fn(() => Promise.resolve());
      const { wrapperInstance } = setup({ editDeck: mockEditDeck });
      jest.spyOn(wrapperInstance, 'setState');

      await wrapperInstance.submitEditedDeck();

      expect(
        wrapperInstance.setState
      ).toHaveBeenCalledWith({ progressDialogVisible: false });
    });

    it('then redirects to MainScreen with a flashMessage', async () => {
      const mockNavigate = jest.fn();
      const mockEditDeck = jest.fn(() => Promise.resolve());
      const { wrapperInstance } = setup({ editDeck: mockEditDeck });
      wrapperInstance.props.navigation.navigate = mockNavigate;

      await wrapperInstance.submitEditedDeck();

      expect(mockNavigate).toHaveBeenCalledWith(
        'Main',
        { flashMessage: 'Deck has been successfully edited' }
      );
    });
  });

  describe('setFormRef', () => {
    it('assigns a ref to this.form', () => {
      const mockForm = jest.fn();
      const { wrapperInstance } = setup();

      wrapperInstance.setFormRef(mockForm);

      expect(wrapperInstance.form).toEqual(mockForm);
    });
  });
});
