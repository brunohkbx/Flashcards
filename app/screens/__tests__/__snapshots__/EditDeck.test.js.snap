// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditDeck renders properly 1`] = `
<Styled(Component)>
  <FormikDeckForm
    formRef={[Function]}
    handleSubmit={[Function]}
    initialValues={
      Object {
        "id": "606a1255",
        "questions": Array [],
        "title": "React",
      }
    }
  />
</Styled(Component)>
`;
